### PR TITLE
fix(custom): fix elements may not be removed after updates #17333

### PR DESCRIPTION
**Pull request by Ovilia**
## Brief Information

This pull request is in the type of:

- [x] bug fixing
- [ ] new feature
- [ ] others

### What does this PR do?

When custom series is updated, if the `renderItem` returns a group that contains null for an element that was previously in the group, it will not be removed, causing some of the elements leave behind by mistake.

### Fixed issues

#17333

### Before: What was the problem?

There is an extra element after calling `setOption` for the second time.

### After: How does it behave after the fixing?

No extra element after calling `setOption` for the second time.

## Document Info

One of the following should be checked.

- [ ] This PR doesn't relate to document changes
- [ ] The document should be updated later
- [x] The document changes have been made in apache/echarts-doc#270

### ZRender Changes

- [ ] This PR depends on ZRender changes (ecomfe/zrender#xxx).

### Related test cases or examples to use the new APIs

- added a test case `custom-update.html`
- tested and passed the visual tests of current cases with "custom" in their names

### Merging options

- [ ] Please squash the commits into a single one when merging.

### CODE REVIEW

**Comment by pissang** [comment]
I'm not pretty sure but I think it's by design at original. Like the logic of option merging, giving `null` means not updating this element in the default merge mode. If we wan't to remove it, we need to change the `$mergeChildren` to false, which will loose the transition. Perhaps @100pah can give more explanation on this.

But from my side. I think the change in this PR will let code much more intuitionistic. So I will give +1 on this change(and a bit more explanation on the doc will be better). The only change I need to request is that we still need to `applyLeaveTransition` when removing the child instead of using `remove` directly

File: src/chart/custom/CustomView.ts [hunk 1]
```diff
@@ -1316,14 +1316,21 @@ function mergeChildren(
...
+                el.childAt(index),
+                dataIndex,
+                newChild as CustomElementOption,
+                seriesModel,
+                el
+            );
+        }
+        else {
+            // The element is being null after updating, remove the old element
+            el.remove(el.childAt(index));
```
Comment (100pah) [inline src/chart/custom/CustomView.ts:1332]: Basically I agree this change, which might make it more intuitive. BTW, using "merge" by default rather than replace because the performance seams to be better.

But this is a breaking change, we need to note it as "breaking change" in release note.

And if modifying this behavior, the comment of this function are also needed to be updated.
> // (1) By default, `elOption.$mergeChildren` is `'byIndex'`, which indicates that
//     the existing children will not be removed, and enables the feature that
//     update some of the props of some of the children simply by construct
//     the returned children of `renderItem` like:
//     `var children = group.children = []; children[3] = {opacity: 0.5};`

Probably need to explain the usage like: `null` means delete and empty object `{}` means keep nothing change.

Do we need to add some doc to explain that in <https://echarts.apache.org/en/option.html#series-custom.renderItem.return_group>?
This doc of `$mergeChildren` is missing because I'm not sure whether to provide this option and whether to name it as `$mergeChildren` yet.

**Review by 100pah — CHANGES_REQUESTED** [review]

File: src/chart/custom/CustomView.ts (hunk 1 above)
Comment (100pah) [reply src/chart/custom/CustomView.ts:1332]:
```js
el.remove(el.childAt(index));
```
This is not correct. `el.remove` will change the entire indices of the subsequent elements

**Review by 100pah — CHANGES_REQUESTED** [review]

**Comment by Ovilia** [comment]
@pissang @100pah Thanks for the review. I updated the comment and add a test case where a child of new group returned from the custom series is `{}`, which should preserve the old element. I will also update the document about this.

File: src/chart/custom/CustomView.ts [hunk 2]
```diff
@@ -1316,24 +1322,43 @@ function mergeChildren(
...
+                api,
+                oldChild,
+                dataIndex,
+                newChild as CustomElementOption,
+                seriesModel,
+                el
+            );
+        }
+        else {
+            removeChildFromGroup(el, oldChild, seriesModel);
```
Comment (100pah) [inline src/chart/custom/CustomView.ts:1338]: If leave animation is disabled, it will still call `parent.remove(el)` directly, which uses `splice` internally and brings a weird behavior:

existing children: `[elA, elB, elC]`
new children: `[elM, null, elN]`
result: `[merge(elA, elM), elC, elN]`


I've changed my opinion that
> Basically I agree this change, which might make it more intuitive

When setOption, the default strategy is "merge", which is the same as the other series like line, scatter, ...
The merge strategy in custom series contains: 
+ dataItems are merged by index
+ group children are merge by index

In series link line, scatter, graphics are the same between different data items, and it makes "merge" strategy make sense in most cases. But is custom series, graphics may not be the same between different data items, which probably make the "merge" not reasonable. 

This change brings a breaking change but only change the `null` case in group children but not considered the case like "merge graphic unexpectedly of different data item", "remain properties unexpectedly when merge graphic". I think it is not good enough.

Could we provide a option on custom series like 
```js
series: {
     type: 'custom',
     merge: false,
}
```
When `merge: false`, there is no merge either between different data items (see https://github.com/apache/echarts/blob/5.3.3/src/chart/custom/CustomView.ts#L222)
or in group children.

And it may fix #17349 thoroughly ?

@Ovilia @pissang what's your opinion?

**Review by 100pah — CHANGES_REQUESTED** [review]

File: src/chart/custom/CustomView.ts (hunk 2 above)
Comment (Ovilia) [reply src/chart/custom/CustomView.ts:1338]: I think this mainly depends on how we understand *merging*. I don't think it's intuitive to use the old element in a group when provided `null` in new `setOption`. This is just like when providing series data like `[10, 20, null, 40]` in a new `setOption` should not take the null data to use the old value in the previous data.

Comment (Ovilia) [reply src/chart/custom/CustomView.ts:1338]: You are quite right! I didn't think of the situation. I fixed this and also update the test case.

Comment (100pah) [reply src/chart/custom/CustomView.ts:1344]: After adding `oldRemovedCount` the effect is better but still has some issues:

## Consider the case below:
+ step 1. `setOption` with new children: `[elA, elB, elC]` (length: 3)
+ step 2. `setOption` with new children: `[elA, null, elC]` (length: 3) (means hide `elB`)
+ step 3. `setOption` with new children: `[elA, elB, elC]` (length: 3) (means show `elB` again)

But with current modified code, the result will be:
+ After step 1. children will be `[elA, elB, elC]`
+ After step 2. children will be `[elA, elC]`
+ After step 3. children will be `[elA, merge(elC, elB), elC]`

which is unexpected.

## Consider leave animation
If the `setOption` is triggered by user's repeat click, the interval between two clicks is not predicable, may be longer or shorter than the leave animation duration. In this case, if the indices of some children is different before and after leave animation, the merge result will be not predicable.

---

Basically, if we use `null/undefined` to delete a child, I think this principle should be apply: 
The index of any other elements should not be changed after this deletion.
That is, is we delete a child, that place may be kept "no element" or "an placeholder element that can not displayed"

**Review by 100pah — CHANGES_REQUESTED** [review]

File: src/chart/custom/CustomView.ts [hunk 3]
```diff
@@ -1316,24 +1322,52 @@ function mergeChildren(
...
-            el.childAt(index),
-            dataIndex,
-            newChildren[index] as CustomElementOption,
-            seriesModel,
-            el
-        );
+        const newChild = newChildren[index];
+        const oldChild = el.childAt(index);
+        if (newChild) {
+            if (newChild.ignore == null) {
```
Comment (100pah) [inline src/chart/custom/CustomView.ts:1328]: Should be:
```ts
if (newChild.ignore) {
    newChild.ignore = false;
}
```
or 
```ts
newChild.ignore = null; // or false
```

**Review by 100pah — CHANGES_REQUESTED** [review]

File: src/chart/custom/CustomView.ts (hunk 3 above)
Comment (Ovilia) [reply src/chart/custom/CustomView.ts:1328]: `newChild` is the user-provided option. So if `newChild.ignore` is true, it means the user want to ignore the element so I don't think we should set `newChild.ignore = false` in this situation.

Comment (100pah) [reply src/chart/custom/CustomView.ts:1328]: I got it wrong

**Review by 100pah — APPROVED** [review]

File: src/chart/custom/CustomView.ts [hunk 4]
```diff
@@ -99,6 +99,7 @@ import {
     applyKeyframeAnimation,
     stopPreviousKeyframeAnimationAndRestore
 } from '../../animation/customGraphicKeyframeAnimation';
+import { SeriesModel } from '../../echarts.all';
```
Comment (plainheart) [inline src/chart/custom/CustomView.ts:102]: The import source is incorrect.

File: src/chart/custom/CustomView.ts (hunk 3 above)
Comment (100pah) [reply src/chart/custom/CustomView.ts:1328]: What if users set a child as `null` / `undefined` at the first time ? (rather than the second time)
